### PR TITLE
Toggle containers programmatically

### PR DESCRIPTION
**Pull request by petrsimon**
Simply add `collapsed` attribute to the `ui-layout-container` you want to control and toggle the value.
```javascript
$scope.layout = {
    toolbar: false,
    statusbar: true
}
```

```html
<div ui-layout-container data-collapsed="layout.toolbar">TOOLBAR</div>
<div ui-layout-container data-collapsed="layout.statusbar">STATUSBAR</div>
```

### CODE REVIEW

**Comment by BobbieBarker** [comment]
Going to need a more extensive code review tomarrow. This pr needs unit tests as well.

**Comment by BobbieBarker** [comment]
Also the commit history needs cleaned up

**Comment by petrsimon** [comment]
Cleaned up the commits. Please, have a look and let me know if this is a way to go. I will then write the unit tests.

**Comment by SomeKittens** [comment]
Thanks, @petrsimon!  This is an oft-requested feature.  I'll let @BobbieBarker do some review, then tests, then bring this in!  Glad to have it.

File: src/ui-layout.js [hunk 1]
```diff
@@ -41,7 +41,8 @@ angular.module('ui.layout', [])
     opts.sizes = opts.sizes || [];
     opts.maxSizes = opts.maxSizes || [];
     opts.minSizes = opts.minSizes || [];
-    opts.dividerSize = opts.dividerSize || 10; //default divider size set to 10
+
+    opts.dividerSize = opts.dividerSize === undefined ? 10 : opts.dividerSize; //default divider size set to 10
```
Comment (SomeKittens) [inline src/ui-layout.js]: Is this related to this commit in particular?  If not, it should be part of a separate PR.

File: src/ui-layout.js [hunk 2]
```diff
@@ -796,6 +807,31 @@ angular.module('ui.layout', [])
...
+                  splitter = ctrl.containers[index - 1];
+                  el = splitter.element[0].children[1];
+                }
+
+
+                $timeout(function(){
+                  angular.element(el).trigger('click');
+                });
+
+                $timeout(function(){
```
Comment (BobbieBarker) [inline src/ui-layout.js]: In this context it looks like $timeout is being used to handle async control flow, would it be possible to rewrite this using a promise/$q. 
Other than that the PR looks good to me.

File: src/ui-layout.js (hunk 1 above)
Comment (petrsimon) [inline src/ui-layout.js]: You are right, not really related. I forgot to take that out. Will PR this separately.

File: src/ui-layout.js [hunk 3]
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
             if(!element.hasClass('stretch')) element.addClass('stretch');
             if(!element.hasClass('ui-layout-container')) element.addClass('ui-layout-container');
 
+            scope.$watch('collapsed', function (val, old) {
+              if (old !== undefined && val !== old) {
```
Comment (wesleycho) [inline src/ui-layout.js]: This is better as `if (angular.isDefined(old)) {` - `val` is guaranteed to not equal `old` by how $watch works.

File: src/ui-layout.js [hunk 4]
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
...
+                } else {
+                  splitter = ctrl.containers[index - 1];
+                  el = splitter.element[0].children[1];
+                  beforeContainer = ctrl.containers[index - 2];
+                  afterContainer = scope.container;
+                }
+
+
+                $timeout(function(){
+                  angular.element(el).trigger('click');
```
Comment (wesleycho) [inline src/ui-layout.js]: This looks like a hack...

Comment (petrsimon) [inline src/ui-layout.js]: Does it ;). I've improved this in the new commit.

File: src/ui-layout.js (hunk 3 above)
Comment (petrsimon) [inline src/ui-layout.js]: Actually during initialization, the equality can actually happen, which would in turn toggle the container against expectations.

**Comment by petrsimon** [comment]
I've made few improvements, please see the new commit.

File: src/ui-layout.js [hunk 5]
```diff
@@ -796,6 +807,37 @@ angular.module('ui.layout', [])
...
+
+                } else {
+                  splitter = ctrl.containers[index - 1];
+                  el = splitter.element[0].children[1];
+                  beforeContainer = ctrl.containers[index - 2];
+                  afterContainer = scope.container;
+                }
+
+
+                $timeout(function(){
```
Comment (BobbieBarker) [inline src/ui-layout.js]: I hate to be a stickler but I was thinking more about using $q instead of $timeout. In my opinion $timeout isn't really appropriate here.

Comment (petrsimon) [inline src/ui-layout.js]: Aha. This is the trick I know how to get around "$digest in progess" error, which was thrown after trying to trigger the click. Not sure how to use $q here.

Comment (petrsimon) [inline src/ui-layout.js]: Btw, why it $timeout not appropriate. I thought this is a common way how to get around $digest error, to execute a routine in another digest cycle with $timeout.

**Comment by petrsimon** [comment]
I've cleaned up few bits, improved it a bit and added few test and demo. Please check it and let me know how it looks now. Will squash the commits later when it's ready for pull.

File: src/ui-layout.js [hunk 6]
```diff
@@ -239,14 +240,14 @@ angular.module('ui.layout', [])
...
-            var child = ctrl.containers[i].element;
+            var container = ctrl.containers[i],
+              child = container.element;
             opts.maxSizes[i] = child.attr('max-size') || child.attr('data-max-size') || opts.maxSizes[i] || null;
             opts.minSizes[i] = child.attr('min-size') || child.attr('data-min-size') || opts.minSizes[i] || null;
-            opts.sizes[i] = child.attr('size') || child.attr('data-size') || opts.sizes[i] || 'auto';
-            //opts.collapsed[i] = child.attr('collapsed') || opts.collapsed[i] || false;
-
 
+            opts.sizes[i] = child.attr('size') || child.attr('data-size') || opts.sizes[i];
```
Comment (SomeKittens) [inline src/ui-layout.js]: Why drop `auto` as an option?

Comment (petrsimon) [inline src/ui-layout.js]: Because the next line takes care of it in case there's no size provided by the user. Am I missing something?

**Comment by petrsimon** [comment]
I've ran into a resizing issues with collapsed containers, which lead me to https://github.com/petrsimon/ui-layout/tree/refactor-display-update (also announced in #100, which it is solution to. Will wait for comments, especially @j--w who was working on it).
I've already merged that branch with this one in https://github.com/petrsimon/ui-layout/tree/toggle-and-resize, which seem to complement each other quite nicely. 
Not sure if I should push it over this one or create new PR from `toggle-and-resize?

**Comment by petrsimon** [comment]
I've made quite few improvements in the toggle-and-resize branch and it all seems to be behaving quite nicely. I had to cut a bit deeper to get around bugs that appeared and I'm not sure if I should force push it here or rather create new PR. Please see b4eb6c2. I will cool off for a while and stop working on it, unless I run into bugs in my application.

**Comment by j--w** [comment]
@petrsimon I made some comments on #100. I would think that tidying up the toggle-and-resize branch and creating a new PR would be the best course of action.

**Comment by petrsimon** [comment]
@j--w, thanks. I thought new PR might be better. Closing this in favour of #154.
